requestStorageAccessForOrigin.bs

```diff
@@ -0,0 +1,213 @@
+
+<h3 id="the-document-object">Changes to {{Document}}</h3>
+
+<pre class="idl">
+partial interface Document {
+  Promise&lt;undefined> requestStorageAccessForOrigin(USVString origin);
+};
+</pre>
+
+When invoked on {{Document}} |doc| with {{USVString}} |requestedOrigin|, the <dfn export method for=Document><code>requestStorageAccessForOrigin(requestedOrigin)</code></dfn> method must run these steps:
+
+1. Let |p| be [=a new promise=].
+1. If |doc| is not [=Document/fully active=], then [=reject=] |p| with an "{{InvalidStateError}}" {{DOMException}} and return |p|.
+1. If |doc|'s [=node navigable=] is not a [=traversable navigable=], [=reject=] |p| with an "{{InvalidStateError}}" {{DOMException}} and return |p|.
+1. If |doc|'s [=Document/origin=] is an [=opaque origin=], [=reject=] |p| with an "{{InvalidStateError}}" {{DOMException}} and return |p|.
+1. Let |parsedURL| be the the result of running the [=URL parser=] on |requestedOrigin|.
+1. If |parsedURL| is failure, [=reject=] |p| with a "{{TypeError}}" {{DOMException}} and return |p|.
+1. Let |origin| be |parsedURL|'s [=origin=].
+1. If |origin| is an [=opaque origin=], [=reject=] |p| with an "{{InvalidStateError}}" {{DOMException}} and return |p|.
+1. If |doc|'s [=Document/origin=] is [=same origin=] with |origin|, [=resolve=] and return |p|.
+1. Let |descriptor| be a newly created {{TopLevelStorageAccessPermissionDescriptor}} with <code>name</code> set to "<a permission><code>top-level-storage-access</code></a>" and with <code>requestedOrigin</code> set to |origin|.
+1. Let |global| be |doc|'s [=relevant global object=].
+1. Run these steps [=in parallel=]:
+    1. Let |hasAccess| be [=a new promise=].
+    1. Let |top-level site| be the result of [=obtain a site|obtaining a site=] from |doc|'s [=relevant settings object=]'s [=environment settings object/origin=].
+    1. [=Determine the top-level storage access policy=] with |descriptor|, |doc|, |top-level site|, and |hasAccess|.
+    1. [=Queue a global task=] on the [=permission task source=] given |global| to resolve or reject |p| based on the result of |hasAccess|.
+1. Return |p|.
+
+<h3 id="ua-policies">User Agent top-level storage access policies</h3>
+
+To <dfn type="abstract-op">determine if a request has top-level storage access</dfn> with [=request=] |request|, run these steps:
+
+ISSUE: should all cross-site subresource requests behave this way, or only those initiated by the top-level document, or top-level plus those requested from an iframe that is same-site with the top-level document, or only those that have explicitly requested access (in which case a set of requested origins would be maintained)?
+
+<!-- this long chain pulled from fetch, but may be unusual or unsafe -->
+1. Let |settings| be |request|'s [=request/client=]'s [=relevant global object=]'s [=associated document=]'s [=relevant settings object=].
+1. Let |embedded origin| be |request|'s [=request/url=].
```

Comment:
This variable is of type "URL" but the variable name is "origin". That's a little confusing, but I guess it matches the fact that TopLevelStorageAccessPermissionDescriptor's requestedOrigin is a USVString (not an origin), but I think we still need to pull the _serialized_ origin out of the URL right? Or else the dictionary's requestedOrigin member will actually be the entire URL string?

Reply:
hm...I went with USVString for the descriptor, but perhaps it would be better to use origin? I wasn't sure what the correct IDL type would be. Regardless, yes, I at least switched to the request's url's origin here.